tools/awslint/GUIDELINES.md

```diff
@@ -0,0 +1,295 @@
+Sometimes they might also represent a **list of strings**.
+
+Since we want
+attributes can either be late-bound ("a promise to a string") or concrete ("a
+string"), the AWS CDK has a mechanism called "tokens" which allows codifying
+late-bound values into strings or string arrays. This approach was chosen in
+order to dramatically simplify the type-system and ergonomics of CDK code. As
+long as users treat these attributes as _opaque values_ (e.g. not try to parse
+them or manipulate them), they can be used
+
+
+One of the design tenets of the AWS CDK is to be strongly-typed.
+
+- Runtime attributes should be represented as readonly properties on the resource class.
+- Runtime attributes should use a the `string` type if the attribute resolves
+
+## Resource Properties
+
+
+
+## Internal/External Polymorphism
+
+
+
+
+The following patterns should be implemented by all resource classes in the AWS
+Construct Library.
+
+Terminology:
+- "Xxx": represents the CloudFormation resource name ().
+
+Let's start by defining an interface which represents a resource of the type
+"Xxx" (e.g. `Bucket`, `Topic`, `Function`, etc).
+
+This interface should represent both resources defined within the same stack
+(aka "internal", "owned") and existing resources that are defined in another
+stack/app (aka "imported", "existing", "external" or "unowned"). Throughout this
+document we shall refer to these two types of resources as **"internal"** and
+**"external"**. Therefore, this is the recommended type to use in public APIs
+that require this resource.
+
+[test](#resource-class-is-construct)
+
+The interface should include all methods and properties, and extend any other
+interfaces that are applicable for both internal and external resources of this
+type.
+
+Here's a template:
+
+```ts
+// extends all interfaces that are applicable for both internal
+// and external resources of this type
+export interface IXxx extends IFoo, IGoo {
+
+  // attributes
+  readonly xxxArn: string;
+  readonly xxxFoo: string;
+  readonly xxxBar: string;
+
+  // security group connections (if applicable)
+  readonly connections: ec2.Connections;
+
+  // permission grants (adds statements to the principal's policy)
+  grant(principal?: iam.IPrincipal, ...actions: string[]): void;
+  grantFoo(principal?: iam.IPrincipal): void;
+  grantBar(principal?: iam.IPrincipal): void;
+
+  // resource policy (if applicable)
+  addToResourcePolicy(statement: iam.PolicyStatement): void;
+
+  // role (if applicable)
+  addToRolePolicy(statement: iam.PolicyStatement): void;
+
+  // pipeline (if applicable)
+  addToPipeline(stage: pipelineapi.IStage, name: string, props?: XxxActionProps): XxxAction;
+
+  // metrics
+  metric(metricName: string, props?: cloudwatch.MetricCustomization): cloudwatch.Metric;
+  metricFoo(props?: cloudwatch.MetricCustomization): cloudwatch.Metric;
+  metricBar(props?: cloudwatch.MetricCustomization): cloudwatch.Metric;
+
+  // export
+  export(): XxxAttributes;
```

Comment:
Don't think that `export()` needs to be on `IXxx`, it could just be on `Xxx`.

Reply:
I think it does need to be on the interface because there are situations where you want to create "composite" exports and call `.export()` on one of your inputs:

```ts
interface MyCompositeProps {
  bucket: s3.IBucket;
  topic: sns.ITopic;
}

class MyComposite extends cdk.Construct {
  public export(): MyCompositeImportProps {
    return {
      bucket: this.bucket.export(),
      topic: this.topic.export()
    }
  }
}
```

In this scenario, you don't know if the `bucket` or `topic` provided to you are internal or external resources, but you still want export to work.